src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetryConsumeResultExtensions.cs

```diff
@@ -0,0 +1,161 @@
+#endif
+
+        try
+        {
+            propagationContext = ExtractPropagationContext(consumeResult.Message?.Headers);
+            return true;
+        }
+        catch
+        {
+            propagationContext = default;
+            return false;
+        }
+    }
+
+    /// <summary>
+    /// Consumes a message and creates <see href="https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#span-kind">a process span</see> embracing the <see cref="ConsumeAndProcessMessageHandler{TKey,TValue}"/>.
+    /// </summary>
+    /// <param name="consumer">The <see cref="IConsumer{TKey,TValue}"/>.</param>
+    /// <param name="handler">A <see cref="ConsumeAndProcessMessageHandler{TKey,TValue}"/>.</param>
+    /// <param name="cancellationToken">An optional <see cref="CancellationToken"/>.</param>
+    /// <typeparam name="TKey">The type of key of the <see cref="ConsumeResult{TKey,TValue}"/>.</typeparam>
+    /// <typeparam name="TValue">The type of value of the <see cref="ConsumeResult{TKey,TValue}"/>.</typeparam>
+    /// <returns>A <see cref="ValueTask"/>.</returns>
+    public static async ValueTask<ConsumeResult<TKey, TValue>?> ConsumeAndProcessMessageAsync<TKey, TValue>(
+        this IConsumer<TKey, TValue> consumer,
+        ConsumeAndProcessMessageHandler<TKey, TValue> handler,
+        CancellationToken cancellationToken = default)
+    {
+#if NETFRAMEWORK
+        if (consumer == null)
+        {
+            throw new ArgumentNullException(nameof(consumer));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(consumer);
+#endif
+
+        if (consumer is not InstrumentedConsumer<TKey, TValue> instrumentedConsumer)
+        {
+            throw new ArgumentException("Invalid consumer type.", nameof(consumer));
+        }
+
+#if NETFRAMEWORK
+        if (handler is null)
+        {
+            throw new ArgumentNullException(nameof(handler));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(handler);
+#endif
+
+        DateTimeOffset start = DateTimeOffset.UtcNow;
+
+        var consumeResult = instrumentedConsumer.Consume(cancellationToken);
+
+        if (consumeResult?.Message == null || consumeResult.IsPartitionEOF)
+        {
+            return consumeResult;
+        }
+
+        Activity? processActivity = null;
+        if (TryExtractPropagationContext(consumeResult, out var propagationContext))
+        {
+            TKey key = consumeResult.Message.Key;
+            object? keyAsObject = key;
+            processActivity = StartProcessActivity(
+                propagationContext,
+                start,
+                consumeResult.TopicPartitionOffset,
+                keyAsObject,
+                instrumentedConsumer.Name,
+                instrumentedConsumer.GroupId!);
+        }
+
+        try
+        {
+            await handler(consumeResult, cancellationToken).ConfigureAwait(false);
+        }
+        finally
```

Comment:
* If handler throws would we want to flag the `Activity` as error status? Also if user wants to set the status of the span would we expect them to do that through `Activity.Current`? Would it be cleaner if we passed in `processActivity` to the callback?

* Is there any kind of duration metric defined for processing a message in the semantic conventions? Just wondering if we should tick a histogram here or anything as well.

Reply:
we do need to set status to error and also report `error.type` attribute.

And yep, messaging semconv define metrics. We've revised them recently and it's best to add metrics later (once next semconv version is released or after we freeze conventions which should happen relatively soon).

Reply:
Updated with your suggestions. 
- When the handler throws ths activity status is set to error, the attribute error.type is set valued with the exception type (i attempted here to limit the cardinality).
- The handler delegate type is enriched with the activity we may have created.
- Regarding metric we already implement metrics from 1.24.0 https://github.com/open-telemetry/semantic-conventions/blob/v1.24.0/docs/messaging/messaging-metrics.md and those metrics will be reported.

Reply:
In 1.24.0 we have `[messaging.deliver.duration](https://github.com/open-telemetry/semantic-conventions/blob/v1.24.0/docs/messaging/messaging-metrics.md#metric-messagingdeliverduration)` that should cover the same duration as process span. 

Still, it'll be renamed and I'd consider waiting for stable messaging semconv to implement it.

Reply:
Then i will implement this new metric in a PR dedicated to 1.24.0 - 1.26.0 migration